From fe5345148cc91cc26e4f92417a83d21e77050656 Mon Sep 17 00:00:00 2001
From: Tomas Winkler <tomas.winkler@intel.com>
Date: Thu, 10 Oct 2019 16:23:47 +0300
Subject: [PATCH 015/122] mei: dal: fix dal read function

1. Check if the file descriptor is in non blocking
mode and return -EGAIN if data is not ready.

2. Return correct errnos

Signed-off-by: Tomas Winkler <tomas.winkler@intel.com>
---
 drivers/misc/mei/dal/dal_cdev.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/misc/mei/dal/dal_cdev.c b/drivers/misc/mei/dal/dal_cdev.c
index 346a8be4432d..befeb1b3c0cc 100644
--- a/drivers/misc/mei/dal/dal_cdev.c
+++ b/drivers/misc/mei/dal/dal_cdev.c
@@ -110,18 +110,19 @@ static ssize_t dal_dev_read(struct file *fp, char __user *buf,
 	if (!buf)
 		return -EINVAL;
 
+	if (kfifo_is_empty(&dc->read_queue))
+		if (fp->f_flags & O_NONBLOCK)
+			return -EAGAIN;
+
 	ret = dal_wait_for_read(dc);
 	if (ret)
 		return ret;
 
-	if (kfifo_is_empty(&dc->read_queue))
-		return 0;
-
 	r_len = kfifo_out(&dc->read_queue, &len, sizeof(len));
 	if (r_len != sizeof(len) || len > count) {
 		dev_dbg(&ddev->dev, "could not copy buffer: src size = %zd, dest size = %zu\n",
 			len, count);
-		return -EFAULT;
+		return -EMSGSIZE;
 	}
 
 	/**
@@ -133,7 +134,7 @@ static ssize_t dal_dev_read(struct file *fp, char __user *buf,
 	ret = kfifo_to_user(&dc->read_queue, buf, len, &copied);
 	if (ret) {
 		dev_dbg(&ddev->dev, "copy_to_user() failed\n");
-		return -EFAULT;
+		return ret;
 	}
 
 	return copied;
-- 
2.17.1

